src/autoTune0.ml

```diff
@@ -0,0 +1,105 @@
+  Printf.printf "functionCalls: %d\n" f.functionCalls;
+  Printf.printf "loops: %d\n" f.loops;
+  Printf.printf "loopBreaks: %d\n" f.loopBreaks;
+  Printf.printf "controlFlowStatements: %d\n" f.controlFlowStatements;
+  Printf.printf "expressions: %d\n" f.expressions;
+  Printf.printf "instructions: %d\n" f.instructions;
+  Printf.printf "integralVars: (%d,%d)\n" (fst f.integralVars) (snd f.integralVars);
+  Printf.printf "arrayVars: (%d,%d)\n" (fst f.arrayVars) (snd f.arrayVars);
+  Printf.printf "pointerVars: (%d,%d)\n" (fst f.pointerVars) (snd f.pointerVars);
+  flush stdout;
+
+
+class collectComplexityFactorsVisitor(factors) = object
+  inherit nopCilVisitor
+
+  method! vfunc _ = factors.functions <- factors.functions + 1; DoChildren
+
+  method! vvdec var =
+    let incVar (g,l) = if var.vglob then (g + 1, l) else (g, l+1) in
+    (if isIntegralType var.vtype then
+       factors.integralVars <- incVar factors.integralVars
+     else if isPointerType var.vtype then
+       factors.pointerVars <- incVar factors.pointerVars
+     else if isArrayType var.vtype then
+       factors.arrayVars <- incVar factors.arrayVars
+    ); DoChildren
+
+  method! vexpr _ = factors.expressions <- factors.expressions + 1; DoChildren
+
+  method! vinst = function
+    | Set _ ->
+      factors.instructions <- factors.instructions + 1; DoChildren
+    | Call (Some _, _,_,_,_) ->
+      factors.instructions <- factors.instructions + 2; (*Count function call and assignment of the result seperately *)
+      factors.functionCalls <- factors.functionCalls + 1; DoChildren
+    | Call _ ->
+      factors.instructions <- factors.instructions + 1;
+      factors.functionCalls <- factors.functionCalls + 1; DoChildren
+    | _ -> DoChildren
+
+  method! vstmt stmt = match stmt.skind with
+    | Loop _ ->
+      factors.controlFlowStatements <- factors.controlFlowStatements + 1;
+      factors.loops <- factors.loops + 1; DoChildren
+    | If _
+    | Switch _
+    | Goto _
+    | ComputedGoto _
+    | Return _ -> factors.controlFlowStatements <- factors.controlFlowStatements + 1; DoChildren
+    | Break _
+    | Continue _ ->
+      factors.controlFlowStatements <- factors.controlFlowStatements + 1;
+      factors.loopBreaks <- factors.loopBreaks + 1; DoChildren
+    | _ -> DoChildren
+
+end
+
+let collectFactors visitAction visitedObject =
+  let factors = {
+    functions = 0;
+    functionCalls = 0;
+    loops = 0;
+    loopBreaks = 0;
+    controlFlowStatements = 0;
+    expressions = 0;
+    instructions = 0;
+    integralVars = (0,0);
+    arrayVars = (0,0);
+    pointerVars = (0,0);
+  } in
+  let visitor = new collectComplexityFactorsVisitor(factors) in
+  ignore (visitAction visitor visitedObject);
+  factors
+
+
+
+(*array heuristics*)
+let is_important_type (t: typ): bool = match t with
+  | TNamed (info, attr) -> List.mem info.tname ["pthread_mutex_t"; "spinlock_t"; "pthread_t"]
+  | TInt (IInt, attr) -> hasAttribute "mutex" attr
+  | _ -> false
```

Comment:
Important in what sense?

Reply:
Here I meant important for thread analysis, The thought was to unroll arrays containing threads and locks so that the analysis has more information about them.

Reply:
Maybe a comment will help here?